### PR TITLE
Port "Supermatter Engine" from EE

### PR DESCRIPTION
**Pull request by WarMechanic**
This one's been a long time coming. Finally, engineering gets more content.

Ports [the supermatter engine](https://github.com/Simple-Station/Einstein-Engines/pull/341) from EE.
Does not actually map the supermatter engine into the maps that need it.

## Media

image
image

I'll mark the PR as review ready, but I'll just do the videos when I can.
- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

Nope, it works

**Changelog**

VMSolidus, White Dream, Colin-Tel, WarMechanic
- add: Supermatter Engines have been implemented.

### CODE REVIEW

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+[RegisterComponent, NetworkedComponent]
+public sealed partial class SupermatterComponent : Component
+{
+    #region Base
+
+    /// <summary>
+    ///     The SM will only cycle if activated.
+    /// </summary>
+    [DataField]
+    public bool Activated = false;
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:19]: dont need = false

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+    [DataField]
+    public string SliverPrototype = "SupermatterSliver";
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:22]: EntProtoId

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+    /// <summary>
+    ///     Affects delamination timer.
+    ///     If removed - delamination timer is divided by 2.
+    /// </summary>
+    [DataField]
+    public bool SliverRemoved = false;
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:29]: dont need = false

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+    public string[] LightningPrototypes =
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:31]: add datafield and use EntProtoId

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    {
+        "Lightning",
+        "ChargedLightning",
+        "SuperchargedLightning",
+        "HyperchargedLightning"
+    };
+
+    [DataField]
+    public string SingularitySpawnPrototype = "Singularity";
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:40]: entprotoid for all these...

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    public float DamageEmergencyThreshold = 500;
+
+    /// <summary>
+    ///     The point at which the SM begins delaminating.
+    /// </summary>
+    [DataField]
+    public int DamageDelaminationPoint = 900;
+
+    [DataField]
+    public bool Delamming = false;
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:292]: dont need

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+    #region Announcements
+
+    [DataField]
+    public string AlertCodeYellowId = "yellow";
+
+    [DataField]
+    public string AlertCodeDeltaId = "delta";
+
+    [DataField]
+    public bool DelamAnnounced = false;
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:308]: dont need

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+
+    #endregion
+
+    #region Gases
+
+    /// <summary>
+    ///     How much gas is in the SM
+    /// </summary>
+    [DataField]
+    public Dictionary<Gas, float> GasStorage = new Dictionary<Gas, float>()
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:318]: should be using a GasMixture

File: Content.Shared/Supermatter/Components/SupermatterComponent.cs [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+        { Gas.Frezon,        0f },
+        { Gas.Ammonia,       0f },
+        { Gas.NitrousOxide,  0f },
+    };
+
+    /// <summary>
+    ///     Stores information about how every gas interacts with the SM
+    /// </summary>
+    //TODO: Replace this with serializable GasFact array something
+    public readonly Dictionary<Gas, (float TransmitModifier, float HeatPenalty, float PowerMixRatio)> GasDataFields = new()
```
Comment (deltanedas) [inline Content.Shared/Supermatter/Components/SupermatterComponent.cs:335]: GasFact exists

**Review by deltanedas — CHANGES_REQUESTED** [review]
all these files should be in ee or deltav folders

**Comment by DangerRevolution** [comment]
THE GUIDEBOOK ENTRY NEEDS A TITLE STRING!!

**Comment by VMSolidus** [comment]
> all these files should be in ee or deltav folders

Any changes you wish made to this PR should be made as a PR to the original source.

**Comment by VMSolidus** [comment]
> all these files should be in ee or deltav folders

Make a PR of these to EE, the original source for the code here. I'll not have DV use an entire different version of Supermatter when it is maintained elsewhere.

**Comment by Samsterious** [comment]
One of the problems we are faced with in Delta-V that is not a problem on Wizden is that teslaloose/singuloose/delam is not a fun or engaging mechanic for 99% of the station's crew. It is sudden death with zero agency. One engineer's mistake forces round end for the entire station. In an LRP context, "funni" is an adequate justification to blow up the station. In MRP, that does not suffice.

I definitely think we want supermatter engines, if only because they sound so fucking cool, but I think we should dedicate some time to thinking about how we can make these scenarios more engaging or whether we just want to make these engines fail gracefully (i.e. it simply stops working or maybe just blows up the chamber instead of killing half the people on the station).

**Comment by Bonktrauma** [comment]
> One of the problems we are faced with in Delta-V that is not a problem on Wizden is that teslaloose/singuloose/delam is not a fun or engaging mechanic for 99% of the station's crew. It is sudden death with zero agency. One engineer's mistake forces round end for the entire station. In an LRP context, "funni" is an adequate justification to blow up the station. In MRP, that does not suffice.


The thing is, You KNOW if it's going to delam, it warns you, ALOT and usually it's rather slow, so like singularity, aslong as you set it up properly, you should be fine.
If one engineer messes up, usually you'll know

**Comment by TadJohnson00** [comment]
Is this still being worked on @WarMechanic?

**Comment by MilonPL** [comment]
Since this PR has been set to draft and received no commits since being created 3 months ago, I will be closing it for now. If you ever return to this PR, feel free to open up a new submission.

**Pull request closed without merging**